pkg/s3select/csv/reader.go

```diff
@@ -145,59 +100,191 @@ func (r *Reader) Read(dst sql.Record) (sql.Record, error) {
 
 // Close - closes underlying reader.
 func (r *Reader) Close() error {
+	if r.close != nil {
+		close(r.close)
+		r.close = nil
+	}
 	return r.readCloser.Close()
 }
 
+// nextSplit will attempt to skip a number of bytes and
+// return the buffer until the next newline occurs.
+// The last block will be sent along with an io.EOF.
+func (r *Reader) nextSplit(skip int, dst []byte) ([]byte, error) {
+	if cap(dst) < skip {
+		dst = make([]byte, 0, skip+1024)
+	}
+	dst = dst[:skip]
+	if skip > 0 {
+		n, err := io.ReadFull(r.buf, dst)
+		if err != nil && err != io.ErrUnexpectedEOF {
+			// If an EOF happens after reading some but not all the bytes,
+			// ReadFull returns ErrUnexpectedEOF.
+			return nil, err
+		}
+		dst = dst[:n]
+	}
+	// Read until next line.
+	in, err := r.buf.ReadBytes('\n')
+	dst = append(dst, in...)
+	return dst, err
+}
+
+func (r *Reader) startReaders(in io.Reader, newReader func(io.Reader) *csv.Reader) error {
+	if r.args.FileHeaderInfo != none {
+		// Read column names
+		// Get one line.
+		b, err := r.nextSplit(0, nil)
+		if err != nil {
+			r.err = err
+			return err
+		}
+		reader := newReader(bytes.NewReader(b))
+		record, err := reader.Read()
+		if err != nil {
+			r.err = err
+			if err != io.EOF {
+				r.err = errCSVParsingError(err)
+				return errCSVParsingError(err)
+			}
+			return err
+		}
+
+		if r.args.FileHeaderInfo == use {
+			// Copy column names since records will be reused.
+			columns := append(make([]string, 0, len(record)), record...)
+			r.columnNames = columns
+		}
+	}
+	const splitSize = 128 << 10
+
+	r.bufferPool.New = func() interface{} {
+		return make([]byte, splitSize+1024)
+	}
+
+	// Create queue
+	r.queue = make(chan *queueItem, runtime.GOMAXPROCS(0))
+	r.input = make(chan *queueItem, runtime.GOMAXPROCS(0))
```

Comment:
Are you avoiding allocation earlier and allocating later to save memory ? For these channels ?

Reply:
Not 100% sure what you mean. Is it the location of this code.

I do them here because it is close to where they are used - mainly just so you don't have to scroll to see what they are set to.

Reply:
ah okay, no I was wondering if these can be allocated in NewReader() itself rather than here @klauspost

Reply:
They *can*. I can move them, though honestly I like to keep them close to where they are closed - the splitter below.